types/src/crypto/indexed_aead/keys.rs

```diff
@@ -0,0 +1,319 @@
+
+use std::marker::PhantomData;
+
+use crate::{
+    LibraryError,
+    crypto::{
+        ear::{
+            AEAD_KEY_SIZE, EarDecryptable, EarEncryptable, EarKey,
+            keys::{EncryptedUserProfileKey, EncryptedUserProfileKeyCtype, IdentityLinkWrapperKey},
+        },
+        errors::{DecryptionError, EncryptionError, RandomnessError},
+        kdf::{KDF_KEY_SIZE, KdfDerivable, KdfKey},
+        secrets::Secret,
+    },
+    identifiers::QualifiedUserName,
+};
+use serde::{Deserialize, Serialize};
+use sqlx::{
+    Database, Decode, Encode, Sqlite, Type, encode::IsNull, error::BoxDynError,
+    sqlite::SqliteTypeInfo,
+};
+use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};
+use tracing::error;
+
+/// Marker trait for indexed keys
+pub trait IndexedKeyType {
+    type DerivationContext<'a>: tls_codec::Serialize + Clone;
+
+    // The `LABEL` constant is used to identify the key type in the database.
+    const LABEL: &'static str;
+}
+
+/// Marker trait for keys that can be randomly generated
+pub trait RandomlyGeneratable {}
+
+// Dummy wrapper type to avoid orphan problem.
+#[derive(Default)]
+pub struct KeyTypeInstance<KT: IndexedKeyType>(PhantomData<KT>);
+
+impl<'q, KT: IndexedKeyType> Encode<'q, Sqlite> for KeyTypeInstance<KT> {
+    fn encode_by_ref(
+        &self,
+        buf: &mut <Sqlite as Database>::ArgumentBuffer<'q>,
+    ) -> Result<IsNull, BoxDynError> {
+        <&str as Encode<Sqlite>>::encode_by_ref(&KT::LABEL, buf)
+    }
+}
+
+impl<'r, KT: IndexedKeyType> Decode<'r, Sqlite> for KeyTypeInstance<KT> {
+    fn decode(value: <Sqlite as Database>::ValueRef<'r>) -> Result<Self, BoxDynError> {
+        let label: &str = Decode::<Sqlite>::decode(value)?;
+        if label != KT::LABEL {
+            return Err(BoxDynError::from(format!(
+                "Invalid key type label: expected {}, got {}",
+                KT::LABEL,
+                label
+            )));
+        }
+        Ok(Self(PhantomData))
+    }
+}
+
+impl<KT: IndexedKeyType> Type<Sqlite> for KeyTypeInstance<KT> {
+    fn type_info() -> SqliteTypeInfo {
+        <&str as Type<Sqlite>>::type_info()
+    }
+}
+
+impl<KT: IndexedKeyType> KeyTypeInstance<KT> {
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<KT: IndexedKeyType> tls_codec::Size for KeyTypeInstance<KT> {
+    fn tls_serialized_len(&self) -> usize {
+        KT::LABEL.len()
+    }
+}
+
+impl<KT: IndexedKeyType> tls_codec::Serialize for KeyTypeInstance<KT> {
+    fn tls_serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
+        let label = KT::LABEL.as_bytes();
+        let written = writer.write(label)?;
+        Ok(written)
+    }
+}
+
+#[derive(
+    Serialize, Deserialize, Clone, Debug, PartialEq, Eq, TlsDeserializeBytes, TlsSerialize, TlsSize,
+)]
+#[serde(transparent)]
+pub struct TypedSecret<KT, ST, const SIZE: usize> {
+    secret: Secret<SIZE>,
+    _type: PhantomData<(KT, ST)>,
```

Comment:
If I am not mistaken you need to exclude this field from serde (tls actually seems to ignore it). Otherwise, it will be also serialized/deserialized.

Reply:
As discussed, `#[serde(transparent)]` does it already.